File: ruby/bird-count/bird_count.rb
class BirdCount
  @week = [0, 2, 5, 3, 7, 8, 4]

  def self.last_week
    @week
  end

  def initialize(birds_per_day)
    @week = birds_per_day
  end

  def yesterday
    @week[-2]
  end

  def total
    @week.sum
  end

  def busy_days
    @week.select {|day| day > 4}.size
  end

  def day_without_birds?
    @week.include? 0
  end
end

File: elixir/bird-count/lib/bird_count.ex
defmodule BirdCount do
  @spec today(list()) :: number()
  def today([]), do: nil
  def today([h | _]), do: h

  @spec increment_day_count(list()) :: list()
  def increment_day_count([]), do: [1]
  def increment_day_count([h | t]), do: [h + 1 | t]

  @spec has_day_without_birds?(list()) :: boolean()
  def has_day_without_birds?([]), do: false
  def has_day_without_birds?([0 | _]), do: true
  def has_day_without_birds?([_ | t]), do: has_day_without_birds?(t)

  @spec total(list()) :: number()
  def total([]), do: 0
  def total([h | t]), do: h + total(t)

  @spec busy_days(list()) :: number()
  def busy_days([]), do: 0
  def busy_days([h | t]) when h >= 5, do: 1 + busy_days(t)
  def busy_days([_ | t]), do: busy_days(t)
end
